render/pixman/renderer.c

```diff
@@ -0,0 +1,234 @@
+#include "render/pixman.h"
+
+#include <wlr/render/interface.h>
+#include <wlr/util/log.h>
+
+#include <pixman.h>
+#include <wayland-server.h>
+#include <drm_fourcc.h>
+
+#include <stdlib.h>
+#include <assert.h>
+
+static const struct wlr_renderer_impl renderer_impl;
+
+static struct wlr_pixman_renderer *get_renderer(
+		struct wlr_renderer *wlr_renderer) {
+	assert(wlr_renderer->impl == &renderer_impl);
+	return (struct wlr_pixman_renderer *)wlr_renderer;
+}
+
+static struct wlr_pixman_buffer *get_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer;
+	wl_list_for_each(buffer, &renderer->buffers, link) {
+		if (buffer->buffer == wlr_buffer) {
+			return buffer;
+		}
+	}
+	return NULL;
+}
+
+static void handle_destroy_buffer(struct wl_listener *listener, void *data) {
+	struct wlr_pixman_buffer *buffer =
+		wl_container_of(listener, buffer, buffer_destroy);
+	wl_list_remove(&buffer->link);
+	wl_list_remove(&buffer->buffer_destroy.link);
+
+	pixman_image_unref(buffer->image);
+
+	free(buffer);
+}
+
+static struct wlr_pixman_buffer *create_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (buffer == NULL) {
+		wlr_log_errno(WLR_ERROR, "Allocation failed");
+		return NULL;
+	}
+	buffer->buffer = wlr_buffer;
+	buffer->renderer = renderer;
+
+	struct wlr_dmabuf_attributes dmabuf = {0};
+	if (!wlr_buffer_get_dmabuf(wlr_buffer, &dmabuf)) {
+		goto error_buffer;
+	}
+
+	uint32_t format;
+	switch (dmabuf.format) {
+	case DRM_FORMAT_XRGB8888:
+		format = PIXMAN_x8r8g8b8;
+		break;
+	case DRM_FORMAT_ARGB8888:
+		format = PIXMAN_a8r8g8b8;
+		break;
+	default:
+		wlr_log(WLR_ERROR, "Unsupported DRM format 0x%x\n", dmabuf.format);
+		goto error_buffer;
+	}
+
+	// TODO: ensure wlr_buffer is a drm dumb buffer
+	struct wlr_drm_dumb_buffer *drm_dumb_buffer =
```

Comment:
Brain dump about accessing the dumb buffer data from the pixman renderer: https://github.com/swaywm/wlroots/issues/2399#issuecomment-779742867

Reply:
Fixed in #2700